File: spec/src/test/java/co/mscp/DockerHelper.java
package co.mscp;

import java.io.IOException;
import java.net.URL;
import java.util.Map;


public class DockerHelper {

    // --- INNER TYPES --- //

    public static class DockerParams {
        public final String label;
        public final String dir;
        public final Map<Integer, Integer> portMap;

        public DockerParams(String label, String dir, Map<Integer, Integer> portMap) {
            this.label = label;
            this.dir = dir;
            this.portMap = portMap;
        }
    }


    // --- STATIC FIELDS --- //

    private static final String[] DB_ENV = new String[]{"POSTGRES_PASSWORD=medicom"};
    private static final String LOAD_FLAG = "MySQL init process done. Ready for start up.";

    public static final int COMMON_PORT = 3361;
    public static final int FINANCE_PORT = 3362;

    public static final DockerParams COMMON = new DockerParams("sqlbatch-spec-test-common", "/docker/common",
            FluentMap.with(3306, COMMON_PORT));

    public static final DockerParams FINANCE = new DockerParams("sqlbatch-spec-test-finance", "/docker/finance",
            FluentMap.with(3306, FINANCE_PORT));



    // --- STATIC METHODS --- //

    public static void runDocker(DockerParams p) throws IOException, InterruptedException {
        if(!co.mscp.DockerUtil.isRunning(p.label)) {
            co.mscp.DockerUtil.insertLogSeparator();

            URL dockerFile = CommonSteps.class.getResource(p.dir);
            if(dockerFile == null) {
                throw new Error("Docker file could not found at " + p.dir);
            }
            String imageId = co.mscp.DockerUtil.build(dockerFile);
            co.mscp.DockerUtil.run(p.label, imageId, p.portMap, ContainerNetwork.bridge(), DB_ENV);

            co.mscp.DockerUtil.waitForContainerWithLabel(p.label,LOAD_FLAG);
        }
    }


    public static void stopDocker(DockerParams p) throws IOException, InterruptedException {
        if(co.mscp.DockerUtil.isRunning(p.label)) {
            com.uzabase.jharmony.docker.DockerUtil.stop(p.label);
        }
    }


    public static void runAll() throws IOException, InterruptedException {
        runDocker(COMMON);
        runDocker(FINANCE);
    }


    public static void stopAll() throws IOException, InterruptedException {
        stopDocker(COMMON);
        stopDocker(FINANCE);
    }


    public static void main(String[] args) {
        try {
            stopAll();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

}

File: spec/src/test/scala/co/mscp/medicom/appointment/spec/DockerHelper.scala
package co.mscp.medicom.appointment.spec

import java.net.URL

import co.mscp.foundation.scala.common.{DockerParams, DockerUtil}
import co.mscp.mmk2.net.NetworkUtil


object DockerHelper {

  private val DB_PORT = 6451
  private val DB = DockerParams(
    "medicom-db",
    toUrl("/docker/db"),
    "database system is ready to accept connections",
    Map(5432 -> DB_PORT),
    Seq("POSTGRES_PASSWORD=medicom"))

  val BACKEND_PORT = 6452
  private val BACKEND = DockerParams(
    "medicom-backend",
    null,
    "play.core.server.AkkaHttpServer - Listening for HTTP on",
    Map(9000 -> BACKEND_PORT),
    Seq(s"MEDICOM_DB_URL=jdbc:postgresql://$getHostAddress:$DB_PORT/medicom"))


  def getHostAddress: String = NetworkUtil.getHostInet.getHostAddress


  private def toUrl(path: String): URL =
    Option(classOf[DockerHelper].getResource(path))
      .getOrElse(throw new Error("Docker file could not found at " + path))


  def runBackend(): Unit = if(!DockerUtil.isRunning(BACKEND)) {
    DockerUtil.insertLogSeparator()
    DockerUtil.run("docker.mscp.co/medicom/appointment:local-test", BACKEND)
    DockerUtil.waitForContainerWithLabel(BACKEND)
  }


  def runDb(): Unit = if(!DockerUtil.isRunning(DB)) {
    DockerUtil.insertLogSeparator()
    DockerUtil.run(DockerUtil.build(DB), DB)
    DockerUtil.waitForContainerWithLabel(DB)
  }


  def runAll(): Unit = {
    runDb()
    runBackend()
  }


  def stopAll(): Unit = {
    DockerUtil.stop(BACKEND)
    DockerUtil.stop(DB)
  }


  def main(args: Array[String]): Unit = stopAll()
}


class DockerHelper